app/actions/websocket/roles.ts

```diff
@@ -0,0 +1,119 @@
+    if (!database) {
+        return;
+    }
+
+    // only update Role records that exist in the Role Table
+    const role = safeParseJSON(msg.data.role) as Role;
+    const dbRole = await queryRoleById(database.database, role.id);
+    if (!dbRole) {
+        return;
+    }
+
+    database.operator.handleRole({
+        roles: [role],
+        prepareRecordsOnly: false,
+    });
+}
+
+export async function handleUserRoleUpdatedEvent(serverUrl: string, msg: WebSocketMessage): Promise<void> {
+    const database = DatabaseManager.serverDatabases[serverUrl];
+    if (!database) {
+        return;
+    }
+
+    const currentUserId = await queryCurrentUserId(database.database);
+    if (currentUserId !== msg.data.user_id) {
+        return;
+    }
+
+    const modelPromises: Array<Promise<Model[]>> = [];
+
+    // update Role Table if needed
+    const newRoles = await fetchRolesIfNeeded(serverUrl, Array.from(msg.data.roles), true);
+    const preparedRoleModels = getPreparedRoleModels(database, newRoles);
+    modelPromises.push(preparedRoleModels);
+
+    // update User Table record
+    const user = await queryCurrentUser(database.database);
+    if (!user) {
+        return;
+    }
+
+    user.prepareUpdate((u) => {
+        u.roles = msg.data.roles;
+    });
+
+    try {
+        await database.operator.batchRecords([user]);
+    } catch {
+        // eslint-disable-next-line no-console
+        console.log('FAILED TO BATCH CHANGES FOR USER ROLE UPDATED');
+    }
+    const models = await Promise.all(modelPromises);
+    const flattenedModels = models.flat() as Model[];
+    if (flattenedModels?.length > 0) {
+        await database.operator.batchRecords(flattenedModels);
+    }
+}
+
+export async function handleMemberRoleUpdatedEvent(serverUrl: string, msg: WebSocketMessage): Promise<void> {
+    const database = DatabaseManager.serverDatabases[serverUrl];
+    if (!database) {
+        return;
+    }
+
+    const currentUserId = await queryCurrentUserId(database.database);
+
+    const member = safeParseJSON(msg.data.member) as TeamMembership;
+    if (currentUserId !== member.user_id) {
+        return;
+    }
+
+    const modelPromises: Array<Promise<Model[]>> = [];
+
+    // update Role Table if needed
+    const newRoles = await fetchRolesIfNeeded(serverUrl, Array.from(member.roles), true);
+    const preparedRoleModels = getPreparedRoleModels(database, newRoles);
+    modelPromises.push(preparedRoleModels);
+
+    // update MyTeam Table
+    const preparedMyTeam = prepareMyTeams(database.operator, [], [member]);
+    if (preparedMyTeam) {
+        modelPromises.push(preparedMyTeam[2]);
+    }
+
+    const models = await Promise.all(modelPromises);
+    const flattenedModels = models.flat() as Model[];
+    if (flattenedModels?.length > 0) {
+        await database.operator.batchRecords(flattenedModels);
+    }
+}
+
+async function getPreparedRoleModels(database: ServerDatabase, roles: RolesRequest): Promise<Model[]> {
```

Comment:
this function seems totally unnecessary and the name is also sort of weird `get` and `prepare...` functions live in the `queries` folder

Reply:
fixed and removed completely